src/google/protobuf/compiler/csharp/csharp_map_field.cc

```diff
@@ -91,7 +91,7 @@ void MapFieldGenerator::GenerateMembers(io::Printer* printer) {
 void MapFieldGenerator::GenerateMergingCode(io::Printer* printer) {
   printer->Print(
       variables_,
-      "$name$_.Add(other.$name$_);\n");
+      "foreach (var $name$Value_ in other.$name$_) {\n  if (!$name$_.ContainsKey($name$Value_.Key)) {\n    $name$_.Add($name$Value_.Key, $name$Value_.Value);\n  }\n}\n");
```

Comment:
The merging semantics for maps is to override the existing value.

Also, usually we make a line wrap in the generator when you print '\n' for better readability.

There also needs to be a test to catch this behavior change.